tm2py/components/network/highway/highway_assign.py

```diff
@@ -144,16 +152,55 @@ def run(self):
-                    assign(assign_spec, scenario, chart_log_interval=1)
 
+                if iteration == 0 or not run_dynamic_toll:
+                    assign_spec = self._get_assignment_spec(assign_classes)
+                    self._run_sola_traffic_assignment(scenario, assign_spec, chart_log_interval=1)
+                else: #iteration > 0 and run_dynamic_toll 
+                    # run maximum 5 times of dynamic tolling
+                    # break out the loop if no valuetoll need to be updated
+                    for dynamic_toll_iteration in range(1, 6):
+                        assign_spec = self._get_assignment_spec(assign_classes)
+                        if dynamic_toll_iteration < 5:
+                            assign_spec["stopping_criteria"]["max_iterations"] = self.controller.config.highway.tolls.dynamic_toll_inner_iter
+                        stopping_criteria = assign_spec["stopping_criteria"]
+                        self._run_sola_traffic_assignment(scenario, assign_spec, chart_log_interval=1)
+                        self._calc_total_flow(scenario)
+                        self._calc_vc(scenario)
+
+                        # reset indicator variable and attribute
+                        update_toll_required = False
+                        net_calc = NetworkCalculator(scenario)
+                        net_calc("@update_dynamic_toll", "0")
+
+                        # check if valuetoll need to be updated
+                        network = scenario.get_network()
+                        for link in network.links():
+                            if link["@vc"] > 1 and link["@tollbooth"] >= valuetoll_start_tollbooth_code:
+                                for dst_veh in dst_veh_groups:
+                                    if link[f"@valuetoll_{dst_veh}"] < max_dynamic_valuetoll:
+                                        # if any of the valuetoll field meet update criteria, flag the link
+                                        update_toll_required = True
+                                        link["@update_dynamic_toll"] = 1
+                                        break 
+
+                        # if need to update valuetoll, call set_tolls function
+                        if update_toll_required:
+                            PrepareNetwork(controller=self._controller)._set_tolls(network=network, time_period=time)
+                            scenario.publish_network(network)
+                        else:
+                            if dynamic_toll_iteration < 5:
+                                # if there is no valuetoll updated needed before 5th dynamic toll iteration,
+                                # run remaining iterations upto the original max_iteration settings
+                                assign_spec = self._get_assignment_spec(assign_classes)
+                                assign_spec["stopping_criteria"]["max_iterations"] -= self.controller.config.highway.tolls.dynamic_toll_inner_iter
```

Comment:
```suggestion
                                assign_spec["stopping_criteria"]["max_iterations"] -= self.config.tolls.dynamic_toll_inner_iter
```
Can we really reduce the iteration here? Meaning does the assignment keep and restart from the previous state (with the inner iterations)?

Reply:
updated in d65a72d